PubSub: Fix streaming pull incorrectly handling FlowControl max_messages setting

plamut opened the pull request:
Closes #7677.

This PR fixes how subscriber client handles the [FlowControl.max_messages](https://github.com/googleapis/google-cloud-python/blob/master/pubsub/google/cloud/pubsub_v1/types.py#L56) setting. It makes sure that subscriber only delivers (i.e. invokes callbacks) `max_messages` at a time, and resumes invoking callbacks only when the user code acknowledges some of the previously delivered messages.

### How to test

**Steps to reproduce:**
- Prerequsite: Have a working service account, pubsub topic, and subscription created.
- Publish 50 - 100 messages to that topic in batches of random size - [example script](https://github.com/plamut/google-cloud-python/blob/iss-7677-debug-app/pubsub/reproduce_pub.py).
- Run a test subscriber script ([example](https://github.com/plamut/google-cloud-python/blob/iss-7677-debug-app/pubsub/reproduce_sub.py)) that subscribes to the topic using a streaming pull.
- (optional) Apply a [patch](https://gist.github.com/plamut/b6f6851dc68212b382ddc02b877b577e) that adds colors to the log output, makes the latter easier to read.

**Actual result (before the fix):**
- The background consumer does not always get paused in time and pulls an additional batch of messages from the server:
![log_incorrectly_paused](https://user-images.githubusercontent.com/1235663/57545522-dfc52180-735a-11e9-8171-e482cdbf1234.png)
- The client subscriber script often receives too many messages simultaneously, i.e. more than `FlowControl.max_messages`.

**Expected result (after the fix):**
- The background consumer always gets paused in time when the leased messages load hits the limit (`max_messages`):
![log_correctly_paused](https://user-images.githubusercontent.com/1235663/57545633-40545e80-735b-11e9-9208-c9fdb7603fb4.png)
- The client subscriber script never has more than `max_messages` pending messages (received, but not yet acknowledged) - the `"pending ACK"` figure in the logs. The streaming pull process waits until some of the messages are ACK-ed before delivering new messages.
- Excessive messages pulled from the server (i.e. those that would cause an overload) are held in an internal buffer, and are released to the client script before any new messages get pulled from the server. The streaming pull process does not temporarily hold any more messages than necessary.

### Footnotes
For `max_messages` settings greater than 10, one might observe that a maximum of 10 (or less) messages are delivered to the client code at once, especially if message callbacks are slow, e.g. by artificially sleeping for prolonged periods of time.

This is caused by the [default callback thread pool](https://github.com/googleapis/google-cloud-python/blob/master/pubsub/google/cloud/pubsub_v1/subscriber/scheduler.py#L67-L73) size, which has `max_workers` set to 10.

This can be mitigated by creating the subscriber client with a custom `scheduler` that uses a thread poll with a higher cap on the worker thread count, or by using [batch callbacks](https://github.com/googleapis/google-cloud-python/issues/4994) for received messages. The latter feature is almost done, and will submit a PR for it soon.

plamut commented:
Rebased because of a merge conflict when #7954 was merged.

plamut commented:
Putting on hold, because the change is a non-trivial one, and we want to make a PubSub release first. We will unblock the PR once done (reviews still welcome, though).

cc: @sduskis

plamut commented on pubsub/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py:
```diff
@@ -210,26 +217,71 @@ def add_close_callback(self, callback):
-            _LOGGER.debug("Did not resume, current load is %s", self.load)
+        """Check the load and held messages and resume the consumer if needed.
+
+        If there are messages held internally to avoid the leaser over load,
```
That's indeed better, updated.

plamut commented on pubsub/tests/system.py:
```diff
@@ -206,6 +208,80 @@ class CallbackError(Exception):
+
+        # now subscribe and do the main part, check for max pending messages
+        total_messages = sum(batch_sizes)
+        flow_control = types.FlowControl(max_messages=5)
```
Indeed, that's the part that keeps track of how many messages at once the client code has.

plamut commented on pubsub/tests/system.py:
```diff
@@ -206,6 +208,80 @@ class CallbackError(Exception):
+        # With total=50, max messages=5, and processing_time=1 this amounts to
+        # 10 seconds (+ overhead), thus a full minute should be more than enough
+        # for the processing to complete. If not, fail the test with a timeout.
+        for _ in six.moves.range(60):
```
That's a good suggestion. I encapsulated the loop into a future on the `callback`, makes this part of the test code cleaner, thanks.

plamut commented on pubsub/tests/system.py:
```diff
@@ -206,6 +208,80 @@ class CallbackError(Exception):
+        )
+
+        # Expected time to process all messages in ideal case:
+        #     total_messages / FlowControl.max_messages * processing_time
```
While mathematically the same, the latter catches the semantics better. `(total_messages / max_messages)` represents the lower bound on the number of batches the client code needs to process, and processing each batch takes approximately `processing_time` (the time to process a single message).

I added parentheses to the formula for a bit more semantic clarity.

plamut commented:
@sduskis Made the changes suggested, the diff can be seen in the [first force push](https://github.com/googleapis/google-cloud-python/compare/b25c427900e0cf5287ebcd7a30bde16de28a45ed..1b8a212c09955f21ce138c54812376cdde191b75).

(the second force push was just rebasing on top of the latest `master`).

tseaver commented on pubsub/google/cloud/pubsub_v1/subscriber/_protocol/leaser.py:
```diff
@@ -64,30 +64,32 @@ def bytes(self):
-                self._bytes += item.byte_size
-            else:
-                _LOGGER.debug("Message %s is already lease managed", item.ack_id)
+        with self._operational_lock:
```
From my read of the code, this lock is normally used to protect access to the `_thread` attrbute, created in `start` and destroyed in `stop`.  The worker thread runs `maintain_leases`, which copies the `_leased_messages` dict before iterating it.  What race are we trying to prevent here?  If `add`, `remove`, and `maintain_leases` need to synchronize access to `_leased_messages`, then perhaps we need a different lock for that?

tseaver replied:
More generally:  we could improve auditability / comprehensibility of the M/T code by adding a comment for each lock created, documenting what attributes it is supposed to guard.

plamut replied:
I had the impression that "operational lock" is meant to synchronize "operations" on the leaser that can be initiated from external threads (e.g. `start` and `stop`), and used it to make the `add` / `remove` operations atomic. If this causes confusion, it might indeed make sense to use a separate lock, e.g. named `add_remove_lock`?

The motivation behind this is the observation in debug logs that `add()` and `remove()` operations can be intertwined if invoked from different threads at just the "right" moment. Both operations modify the internal `_leased_messages` dict and the `_bytes` counter.

The dict is probably fine, because `ack_id` is unique for every message, but I wouldn't say the same for `_bytes`.

tseaver commented on pubsub/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py:
```diff
@@ -116,6 +117,12 @@ def __init__(
+        # A FIFO queue for the messages that have been received from the server,
+        # but not yet added to the lease management (and not sent to user callback),
+        # because the FlowControl limits have been hit.
+        self._messages_on_hold = queue.Queue()
```
Do we need to set a maximum size on the queue?

plamut replied:
We don't - the queue only gets filled in the `_on_response()` callback, and only after the load reaches 1.0. At that point the consumer is already paused and no more messages arrive.

Additionally, the consumer only gets maybe-resumed after clearing the queue first, meaning that the maximum queue size is bounded by the maximum number of messages the server sends in a single response.

tseaver commented on pubsub/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py:
```diff
@@ -211,26 +218,70 @@ def add_close_callback(self, callback):
+
+        If there are currently no messageges on hold, or if the leaser is
+        already overloaded, this method is effectively a no-op.
+        """
```
Maybe add a comment here that the method assumes the caller has acquired the `_pause_resume_lock`?

plamut replied:
Would not hurt, and explicit is better than implicit.

tseaver commented on pubsub/google/cloud/pubsub_v1/subscriber/message.py:
```diff
@@ -208,8 +212,8 @@ def lease(self):
-            This method is called by the constructor, and you should never
-            need to call it manually.
+            By default this method is called by the constructor, and you should
+            never need to call it manually.
```
Is "never" correct here?  What if the application constructs the message with `lease=False`?

plamut replied:
Re-reading it again, it is not - it refers to the default case, but that's not clear from the sentence, will reword.

plamut commented:
@tseaver Addressed the comments, please have another look. The [diff](https://github.com/googleapis/google-cloud-python/compare/7e5b9d778b8d507a634ef6e0b06e924be2d4ddfe..ec6a21d938ce1e09b2dc583ac78dd6ab36dab8a6) is small enough, thus I changed a few commits and force-pushed the changes.

Also removing the `do not merge` label, since PubSub version `0.41.0` has been successfully released in the meantime.
